src/enumerable.cr

```diff
@@ -1646,30 +1646,34 @@ module Enumerable(T)
   # :nodoc:
   def self.zip(main, others : U, &block) forall U
     {% begin %}
+      {% for type, type_index in U %}
+        other{{type_index}} = others[{{type_index}}]
+      {% end %}
+
       # Try to see if we need to create iterators (or treat as iterators)
       # for every element in `others`.
       {% for type, type_index in U %}
-        {% if type < Indexable %}
+        if other{{type_index}}.is_a?(Indexable)
```

Comment:
`case` would be imo tad moar cleaner here:

```cr
case other{{type_index}}
when Indexable
  # ...
when Iterable
  # ...
end
```

ditto below

Reply:
Thanks, fixed. And also did it a bit differently...